modeling/theoretical-simulator.go

```diff
@@ -27,48 +30,142 @@ func (sim TheoreticalSimulator) Simulate(region regionInfo, endpointSlices map[s
+	for zone := range region.zoneDetails {
+		endpointsTrafficLoad[zone] = map[string]float64{}
+		for label, sliceGroup := range endpointSlices {
+			if sliceGroup.Composition[zone].Number == 0 {
+				continue
 			}
+			// calcualte the ratio of the endpoints in the sliceGroup
+			zoneRatioInSG := float64(sliceGroup.Composition[zone].Number) * sliceGroup.Composition[zone].Weight / sliceGroup.numberOfWeightedEndpoints()
+			// zone endpoints traffic load in this sliceGroup = sliceGroup
+			// traffic * zone ratio in this sliceGroup
+			endpointsTrafficLoad[zone][label] = sgTrafficRatio[label] * zoneRatioInSG / float64(sliceGroup.Composition[zone].Number)
 		}
 	}
-	// probability of traffic from a zone to other zones, outgoing zone --
-	// probability of going to other zones
-	zoneEndpoints := map[string]map[string]float64{}
-	for zoneName, endpointsDistribution := range kubeProxy {
-		zoneTotal := 0.0
-		for _, endpoints := range endpointsDistribution {
-			zoneTotal += endpoints
+	// endpoints traffic load deviation
+	// first key: zone the endpoints belong to
+	// second key: sliceGroup the endpoints assigned to
+	endpointsTrafficLoadDeviation := map[string]map[string]float64{}
+	// theoratially, traffic should be distributed equally among all the
+	// endpoints
+	theoretialTrafficLoad := 1.0 / float64(region.totalEndpoints)
+	for zone := range region.zoneDetails {
+		endpointsTrafficLoadDeviation[zone] = map[string]float64{}
+		for label := range endpointsTrafficLoad[zone] {
+			endpointsTrafficLoadDeviation[zone][label] = endpointsTrafficLoad[zone][label]/theoretialTrafficLoad - 1.0
 		}
-		zoneEndpoints[zoneName] = map[string]float64{}
-		for destZoneName, endpoints := range endpointsDistribution {
-			zoneEndpoints[zoneName][destZoneName] = endpoints / zoneTotal
+	}
+
+	// ratio of traffic from a zone to other zones
+	zoneTrafficToZone := map[string]map[string]float64{}
+	for oriZone := range region.zoneDetails {
+		zoneTrafficToZone[oriZone] = map[string]float64{}
+		for label, sliceGroup := range endpointSlices {
+			if sliceGroup.numberOfWeightedEndpoints() == 0.0 {
+				continue
+			}
+			for destZone := range region.zoneDetails {
+				desZoneRatioInSG := float64(sliceGroup.Composition[destZone].Number) * sliceGroup.Composition[destZone].Weight / sliceGroup.numberOfWeightedEndpoints()
+				// traffic oriZone -> desZone: sum(traffic distribution of
+				// oriZone * traffic ratio from oriZone to this sliceGroup *
+				// desZone ratio in this sliceGroup)
+				zoneTrafficToZone[oriZone][destZone] += region.zoneDetails[oriZone].nodesRatio * zoneTrafficToSG[oriZone][label] * desZoneRatioInSG
+			}
 		}
 	}
 
 	// calculate result of one simulation
 	var simResult SimulationResult
-	simResult.TrafficDetail = map[string]Traffic{}
-	simResult.Workload = map[string]float64{}
-	for zoneName, zone := range region.zoneDetails {
+	// traffic distribution details by zone
+	simResult.TrafficDistribution = map[string]zoneTraffic{}
```

Comment:
This has turned into a really long function, one that would be rather difficult to effectively unit test. I think there are lots of opportunities to split this up, but this is one of the big ones - calculating the results of the simulation.

Reply:
Done